tls/s2n_record_read.c

```diff
@@ -34,15 +34,21 @@ int s2n_sslv2_record_header_parse(
         uint8_t *client_protocol_version,
         uint16_t *fragment_length)
 {
-    struct s2n_stuffer *in = &conn->header_in;
+    struct s2n_stuffer *header_in = &conn->header_in;
 
-    S2N_ERROR_IF(s2n_stuffer_data_available(in) < S2N_TLS_RECORD_HEADER_LENGTH, S2N_ERR_BAD_MESSAGE);
+    POSIX_ENSURE(s2n_stuffer_data_available(header_in) >= S2N_TLS_RECORD_HEADER_LENGTH,
+            S2N_ERR_BAD_MESSAGE);
 
-    POSIX_GUARD(s2n_stuffer_read_uint16(in, fragment_length));
+    POSIX_GUARD(s2n_stuffer_read_uint16(header_in, fragment_length));
+    /* The first bit of the length must be set to indicate SSLv2,
+     * so the raw fragment length must be adjusted.
+     */
+    *fragment_length ^= (S2N_TLS_SSLV2_HEADER_FLAG << 8);
```

Comment:
I do prefer the shift+`fragment_length` modification over the blob.

That being said, there's still an assumption that the first bit is set with the XOR. It might be safer to do an `&`

```suggestion
    *fragment_length &= (S2N_TLS_SSLV2_HEADER_FLAG << 8) - 1;
```

or make sure the bit is actually set:
```suggestion
    POSIX_ENSURE(*fragment_length & S2N_TLS_SSLV2_HEADER_FLAG, S2N_ERR_MAD_MESSAGE);
    *fragment_length ^= (S2N_TLS_SSLV2_HEADER_FLAG << 8);
```

Reply:
It took me longer than I like to understand the version that uses &, so I think I'll go with just making sure that the bit is set. It should be anyway, since otherwise the length is actually supposed to be three bytes long :) sslv2 is fun

Reply:
You can also invert the bits as well; not sure if that's any clearer :)

```c
*fragment_length &= ~(S2N_TLS_SSLV2_HEADER_FLAG << 8);
```